cmd/autoretrieve/bitswap/provider.go

```diff
@@ -0,0 +1,247 @@
+package bitswap
+
+import (
+	"context"
+	"sync"
+
+	"github.com/application-research/estuary/cmd/autoretrieve/blocks"
+	"github.com/application-research/estuary/cmd/autoretrieve/filecoin"
+	"github.com/ipfs/go-bitswap/message"
+	bitswap_message_pb "github.com/ipfs/go-bitswap/message/pb"
+	"github.com/ipfs/go-bitswap/network"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-log/v2"
+	"github.com/ipfs/go-peertaskqueue"
+	"github.com/ipfs/go-peertaskqueue/peertask"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	"github.com/libp2p/go-libp2p-kad-dht/fullrt"
+)
+
+var logger = log.Logger("autoretrieve")
+
+const (
+	wantTypeHave  = bitswap_message_pb.Message_Wantlist_Have
+	wantTypeBlock = bitswap_message_pb.Message_Wantlist_Block
+)
+
+type taskQueueTopic uint
+
+const (
+	topicSendBlock taskQueueTopic = iota
+	topicSendHave
+	topicSendDontHave
+)
+
+const targetMessageSize = 16384
+
+type ProviderConfig struct {
+	DataDir string
+}
+
+type Provider struct {
+	config            ProviderConfig
+	network           network.BitSwapNetwork
+	blockManager      *blocks.Manager
+	retriever         *filecoin.Retriever
+	taskQueue         *peertaskqueue.PeerTaskQueue
+	sendWorkerCount   uint
+	sendWorkerCountLk sync.Mutex
+}
+
+func NewProvider(
+	config ProviderConfig,
+	host host.Host,
+	datastore datastore.Batching,
+	blockManager *blocks.Manager,
+	retriever *filecoin.Retriever,
+) (*Provider, error) {
+
+	fullRT, err := fullrt.NewFullRT(host, dht.DefaultPrefix, fullrt.DHTOption(
+		dht.Datastore(datastore),
+		dht.BucketSize(20),
+		dht.BootstrapPeers(dht.GetDefaultBootstrapPeerAddrInfos()...),
+	))
+	if err != nil {
+		return nil, err
+	}
+
+	provider := &Provider{
+		config:          config,
+		network:         network.NewFromIpfsHost(host, fullRT),
+		blockManager:    blockManager,
+		retriever:       retriever,
+		taskQueue:       peertaskqueue.New(),
+		sendWorkerCount: 0,
+	}
+
+	provider.network.SetDelegate(provider)
+
+	return provider, nil
+}
+
+func (provider *Provider) startSend() {
+	provider.sendWorkerCountLk.Lock()
+	// TODO: use a constant or config val
+	if provider.sendWorkerCount < 1 {
```

Comment:
this is only ever spinning up a single worker?

Reply:
just had to add config val for it, that number changes max workers